mlir/include/mlir/Dialect/MQTOpt/IR/WireIterator.h

```diff
@@ -0,0 +1,337 @@
+                                                        mlir::Region* region) {
+    for (mlir::Operation* user : v.getUsers()) {
+      if (user->getParentRegion() == region) {
+        return user;
+      }
+    }
+    return nullptr;
+  }
+
+public:
+  using iterator_category = std::bidirectional_iterator_tag;
+  using difference_type = std::ptrdiff_t;
+  using value_type = mlir::Operation*;
+
+  explicit WireIterator() = default;
+  explicit WireIterator(mlir::Value q, mlir::Region* region)
+      : currOp(q.getDefiningOp()), q(q), region(region) {}
+
+  [[nodiscard]] mlir::Operation* operator*() const {
+    assert(!sentinel && "Dereferencing sentinel iterator");
+    assert(currOp && "Dereferencing null operation");
+    return currOp;
+  }
+
+  [[nodiscard]] mlir::Value qubit() const { return q; }
+
+  WireIterator& operator++() {
+    advanceForward();
+    return *this;
+  }
+
+  WireIterator operator++(int) {
+    auto tmp = *this;
+    ++*this;
+    return tmp;
+  }
+
+  WireIterator& operator--() {
+    advanceBackward();
+    return *this;
+  }
+
+  WireIterator operator--(int) {
+    auto tmp = *this;
+    --*this;
+    return tmp;
+  }
+
+  bool operator==(const WireIterator& other) const {
+    return other.q == q && other.currOp == currOp && other.sentinel == sentinel;
+  }
+
+  bool operator==([[maybe_unused]] std::default_sentinel_t s) const {
+    return sentinel;
+  }
+
+private:
+  void advanceForward() {
+    /// If we are already at the sentinel, there is nothing to do.
+    if (sentinel) {
+      return;
+    }
+
+    /// Find output from input qubit.
+    /// If there is no output qubit, set `sentinel` to true.
+    if (q.getDefiningOp() != currOp) {
+      mlir::TypeSwitch<mlir::Operation*>(currOp)
+          .Case<UnitaryInterface>(
+              [&](UnitaryInterface op) { q = findOutput(op, q); })
+          .Case<AllocQubitOp>([&](AllocQubitOp op) { q = op.getQubit(); })
+          .Case<ResetOp>([&](ResetOp op) { q = op.getOutQubit(); })
+          .Case<MeasureOp>([&](MeasureOp op) { q = op.getOutQubit(); })
+          .Case<mlir::scf::ForOp>(
+              [&](mlir::scf::ForOp op) { q = findResult(op, q); })
+          .Case<mlir::scf::IfOp>(
+              [&](mlir::scf::IfOp op) { q = findResult(op, q); })
+          .Case<DeallocQubitOp, mlir::scf::YieldOp>(
+              [&](auto) { sentinel = true; })
+          .Default([&](mlir::Operation* op) {
+            report_fatal_error("unknown op in def-use chain: " +
+                               op->getName().getStringRef());
+          });
+    }
+
+    /// Find the next operation.
+    /// If it is a sentinel there are no more ops.
+    if (sentinel) {
+      return;
+    }
+
+    /// If there are no more uses, set `sentinel` to true.
+    if (q.use_empty()) {
+      sentinel = true;
+      return;
+    }
+
+    /// Otherwise, search the user in the targeted region.
+    currOp = getUserInRegion(q, getRegion());
```

Comment:
At this point, it is pretty much guaranteed that `q` has `q.hasOneUse() == true`, right? due to the linear typing of qubit values in the Opt dialect.
Couldn't you simply use `currOp = *q.getUsers().begin()` here? I am probably overlooking something here.

Reply:
Thanks for the explanation. That makes a lot more sense now 👍🏻